packages/runtime/src/runtime/navigateApi.tsx

```diff
@@ -0,0 +1,65 @@
+import type {
+  NavigateScreenAction,
+  CustomScope,
+  ScreenContextDefinition,
+  EnsembleStorage,
+} from "@ensembleui/react-framework";
+import { findExpressions, evaluate } from "@ensembleui/react-framework";
+import { isString, cloneDeep, set } from "lodash-es";
+import type { NavigateFunction, Location } from "react-router-dom";
+
+export interface BrowserHistoryDefinition {
+  navigate: NavigateFunction | undefined;
+  location: Location | undefined;
+}
+
+export const browserHistory: BrowserHistoryDefinition = {
+  navigate: undefined,
+  location: undefined,
+};
+
+export const navigateApi = (
+  targetScreen: NavigateScreenAction,
+  screenContext: ScreenContextDefinition,
+  customScope: CustomScope | undefined,
+  storage: EnsembleStorage,
+): void => {
+  const hasOptions = !isString(targetScreen);
+  const screenName = hasOptions ? targetScreen.name : targetScreen;
+
+  // find the matching screen
+  const matchingScreen = screenContext?.app?.screens.find(
+    (s) => s.name.toLowerCase() === screenName.toLowerCase(),
+  );
+
+  if (!matchingScreen) {
+    return;
+  }
+
+  // set additional inputs
+  const inputs =
+    !isString(targetScreen) && targetScreen.inputs
+      ? cloneDeep(targetScreen.inputs)
+      : {};
+
+  if (screenContext) {
+    const expressionMap: string[][] = [];
+    findExpressions(inputs, [], expressionMap);
+    expressionMap.forEach(([path, value]) => {
+      const result = evaluate(screenContext, value, {
+        ensemble: { storage },
+        ...customScope,
+      });
+      set(inputs, path, result);
+    });
+  }
```

Comment:
Does this not throw errors for you in console? This is being evaluated in the context of the executeCode scope so you can only reference the context being passed in. You don't need to evaluate these because they will be automatically dereferenced in the code block

Reply:
No, this is not throwing any errors in the console. also whatever screen name user passed with the ensemble.navigateScreen API, first we need to validate that screen name is the right one and that screen is exists within the app scope. that's why I need to evaluate this. I am passing the hook context from the useExecuteCode block  to navigateApi, in order to access those hooks.

Reply:
Hmm then the matching screen logic can stay, but the input/expression evaluation doesn't need to be here.